src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs

```diff
@@ -87,6 +90,62 @@ private enum MarshalMode
             MarshallingRealObject = 1,
         }
 
+        internal static bool TryGetMarshalOptionsForType(Type type, [NotNullWhen(true)] out JsonRpcProxyOptions? proxyOptions, [NotNullWhen(true)] out JsonRpcTargetOptions? targetOptions)
+        {
+            proxyOptions = null;
+            targetOptions = null;
+            if (type.IsInterface is false)
+            {
+                return false;
+            }
+
+            if (MarshaledTypes.TryGetValue(type, out (JsonRpcProxyOptions ProxyOptions, JsonRpcTargetOptions TargetOptions) options))
+            {
+                proxyOptions = options.ProxyOptions;
+                targetOptions = options.TargetOptions;
+                return true;
+            }
+
+            foreach ((Type implicitlyMarshaledType, JsonRpcProxyOptions typeProxyOptions, JsonRpcTargetOptions typeTargetOptions) in ImplicitlyMarshaledTypes)
+            {
+                if (implicitlyMarshaledType == type ||
+                    (implicitlyMarshaledType.IsGenericTypeDefinition &&
+                     type.IsConstructedGenericType &&
+                     implicitlyMarshaledType == type.GetGenericTypeDefinition()))
+                {
+                    proxyOptions = typeProxyOptions;
+                    targetOptions = typeTargetOptions;
+                    MarshaledTypes.TryAdd(type, (proxyOptions, targetOptions));
+                    return true;
+                }
+            }
+
+            if (type.GetCustomAttribute<RpcMarshalableAttribute>() is not null)
+            {
+                if (typeof(IDisposable).IsAssignableFrom(type) is false)
```

Comment:
@AArnott, should I allow `IAsyncDisposable` or `Microsoft.VisualStudio.Threading.IAsyncDisposable` as an alternative to `IDisposable`?

Reply:
I don't think so. Proxies never implement these async dispose interfaces because Dispose is always transmitted to the server as a notification, so there is nothing for the proxy to wait for. I think the marshaling behavior is that the server object can implement the async disposable interfaces and if it does, StreamJsonRpc will invoke it, but this doesn't manifest on the proxy. As this interface becomes what the proxy implements, it shouldn't derive from the async interface.